src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp

```diff
@@ -3066,235 +3402,138 @@ void StubGenerator::aesgcm_encrypt(Register in, Register len, Register ct, Regis
-  __ evmovdquq(ZTMP14, Address(ct, ghash_pos, Address::times_1, 1 * 64), Assembler::AVX_512bit);
-  __ vpshufb(ZTMP13, ZTMP13, xmm24, Assembler::AVX_512bit);
-  __ vpshufb(ZTMP14, ZTMP14, xmm24, Assembler::AVX_512bit);
-  __ evmovdquq(ZTMP15, Address(avx512_subkeyHtbl, rbx, Address::times_1, 0 * 64), Assembler::AVX_512bit);
-  __ evmovdquq(ZTMP16, Address(avx512_subkeyHtbl, rbx, Address::times_1, 1 * 64), Assembler::AVX_512bit);
-
-  // ghash blocks 0 - 3
-  carrylessMultiply(ZTMP6, ZTMP7, ZTMP8, ZTMP5, ZTMP13, ZTMP15);
-
-  // ghash blocks 4 - 7
-  carrylessMultiply(ZTMP10, ZTMP11, ZTMP12, ZTMP9, ZTMP14, ZTMP16);
-
-  // update sums
-  // ZTMP1 = ZTMP1 + ZTMP5 + ZTMP9
-  // ZTMP2 = ZTMP2 + ZTMP6 + ZTMP10
-  // ZTMP3 = ZTMP3 + ZTMP7 xor ZTMP11
-  // ZTMP4 = ZTMP4 + ZTMP8 xor ZTMP12
-  xorGHASH(ZTMP1, ZTMP2, ZTMP3, ZTMP4, ZTMP5, ZTMP9, ZTMP6, ZTMP10, ZTMP7, ZTMP11, ZTMP8, ZTMP12);
-  __ addl(ghash_pos, 128);
-  __ addl(rbx, 128);
-  __ jmp(LOOP);
 
-  // Integrate ZTMP3/ZTMP4 into ZTMP1 and ZTMP2
-  __ bind(ACCUMULATE);
-  __ evpxorq(ZTMP3, ZTMP3, ZTMP4, Assembler::AVX_512bit);
-  __ vpsrldq(ZTMP7, ZTMP3, 8, Assembler::AVX_512bit);
-  __ vpslldq(ZTMP8, ZTMP3, 8, Assembler::AVX_512bit);
-  __ evpxorq(ZTMP1, ZTMP1, ZTMP7, Assembler::AVX_512bit);
-  __ evpxorq(ZTMP2, ZTMP2, ZTMP8, Assembler::AVX_512bit);
-
-  // Add ZTMP1 and ZTMP2 128 - bit words horizontally
-  vhpxori4x128(ZTMP1, ZTMP11);
-  vhpxori4x128(ZTMP2, ZTMP12);
-  // Load reduction polynomial and compute final reduction
-  __ evmovdquq(ZTMP15, ExternalAddress(ghash_polynomial_reduction_addr()), Assembler::AVX_512bit, rbx /*rscratch*/);
-  vclmul_reduce(AAD_HASHx, ZTMP15, ZTMP1, ZTMP2, ZTMP3, ZTMP4);
-
-  // Pre-increment counter for next operation
-  __ vpaddd(CTR_BLOCKx, CTR_BLOCKx, xmm18, Assembler::AVX_128bit);
-  // Shuffle counter and save the updated value
-  __ vpshufb(CTR_BLOCKx, CTR_BLOCKx, xmm24, Assembler::AVX_512bit);
+  __ movdl(CTR_CHECK, CTR_BLOCKx);
+  __ andl(CTR_CHECK, 255);
+
+  // Reshuffle counter
+  __ vpshufb(CTR_BLOCKx, CTR_BLOCKx, SHUF_MASK, Assembler::AVX_512bit);
+
+  initial_blocks_16_avx512(in, out, ct, pos, key, avx512_subkeyHtbl, CTR_CHECK, rounds, CTR_BLOCKx, AAD_HASHx,  ADDBE_4x4, ADDBE_1234, ADD_1234, SHUF_MASK, stack_offset, false);
+  __ addl(pos, 16 * 16);
+  __ cmpl(len, 32 * 16);
+  __ jcc(Assembler::below, MESG_BELOW_32_BLKS);
+
+  initial_blocks_16_avx512(in, out, ct, pos, key, avx512_subkeyHtbl, CTR_CHECK, rounds, CTR_BLOCKx, AAD_HASHx, ADDBE_4x4, ADDBE_1234, ADD_1234, SHUF_MASK, stack_offset + 16, false);
+  __ addl(pos, 16 * 16);
+  __ subl(len, 32 * 16);
+
+  __ cmpl(len, 32 * 16);
+  __ jcc(Assembler::below, NO_BIG_BLKS);
+
+  __ bind(ENCRYPT_BIG_BLKS_NO_HXOR);
+  __ cmpl(len, 2 * 32 * 16);
+  __ jcc(Assembler::below, ENCRYPT_BIG_NBLKS);
+  ghash16_encrypt_parallel16_avx512(in, out, ct, pos, avx512_subkeyHtbl, CTR_CHECK, rounds, key, true, true, false, false, false, ghashin_offset, aesout_offset, HashKey_32);
```

Comment:
ghash16_encrypt_parallel16_avx512  needs to pass in CTL_BLOCKx, AAD_HASHx, ADDBE_4x4, ADDBE_1234, AAD_HASHx, SHUF_MASK just like we do in initial_blocks_16_avx512. Also the GL and GH needs to be passed in.